### ISSUE TITLE
Paper/Explanation of algorithm used

### ISSUE THREAD

**Issue by mainrs** [tshatrov/ichiran#23]
I'd like to integrate segmentation into an app I wrote. As far as I know, it's not possible to run Lisp on Android.

Is there some kind of documentation that explains how the segmentation works? Even if it takes more time, I'd like to port it over to Kotlin.

**Comment by tshatrov** [comment]
There's a [blog post](https://readevalprint.tumblr.com/post/97467849358/who-needs-graph-theory-anyway) about the segmentation algorithm, but the secret sauce is really the scoring algorithm, which was built in an ad-hoc manner over the years to split sentences as well as possible. The reference implementation (which is Ichiran) is perhaps the only possible way to describe it.

**Comment by tshatrov** [comment]
Yes, this is the suite, it's not particularly thorough, I was mostly including corner cases for the segmentations I wanted to fix.

https://github.com/tshatrov/ichiran/blob/master/tests.lisp

**Comment by mainrs** [comment]
> I'm actually surprised no one has explored these pure-dictionary segmenters more, since they seem much simpler to implement, have 0 training cost, much higher quality output, and more predictable behavior than the CRF trained ones.

I guess the drawback is that you have to manually add a lot of scoring costs for edge cases or exceptions in grammar or otherwise. I do agree that I get more quality output from ichiran than any other system I have tried!

**Comment by mainrs** [comment]
Would a hybrid approach work? Maybe create a NER model who's only task is to derive entities. And use that as some form of input hint to ichiran.
